model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+                mask           : bits(64) = zeros();
+                foreach (j from 0 to (level - 1))
+                  mask = ((mask << sv_params.pte_PPN_j_size_bits) | mask_PPN_j);
+                // Lower-level PPNs must be zero (for aligned superpage)
+                if not((ppns & mask) == zero_extend(0b0)) then
+                  // misaligned superpage mapping
+                  PTW_Failure(PTW_Misaligned(), ext_ptw)
+                else {
+                  // Compose final PA in superpage:
+                  // Superpage PPN + lower VPNs + pagesize_bits page-offset
+                  let ppn = ppns | (vpns_of_va (sv_params, va) & mask);
+                  let pa  = ((ppn << pagesize_bits) | zero_extend(offset_of_va(va)));
+                  PTW_Success(pa, pte, pte_addr, level, global', ext_ptw)
+                }
+              }
+              else {
+                let pa = ((ppns << pagesize_bits) | zero_extend(offset_of_va (va)));
+                PTW_Success(pa, pte, pte_addr, level, global', ext_ptw)
+              }
+          }
+        }
+      }
+    }
+  }
+}
+
+// ****************************************************************
+// Architectural SATP CSR
+
+// PUBLIC: see also riscv_insts_zicsr.sail and other CSR-related files
+register satp : xlenbits
+
+// See riscv_sys_regs.sail for legalize_satp{32,64}().
+// WARNING: those functions legalize Mode but not ASID?
+// PUBLIC: invoked from writeCSR() to fixup WARL fields
+function legalize_satp(a : Architecture,
+                       o : xlenbits,        // previous value of satp
+                       v : xlenbits)        // proposed new value of satp
+                      -> xlenbits = {       // new legal value of satp
+  if sizeof(xlen) == 32 then {
+    // The slice and extend ops below are no-ops when xlen==32,
+    // but appease the type-checker when xlen==64 (when this code is not executed!)
+    let o32      : bits(32) = o[31 .. 0];
+    let v32      : bits(32) = v[31 .. 0];
+    let new_satp : bits(32) = legalize_satp32(a, o32, v32);
+    zero_extend(new_satp);
+  } else if sizeof(xlen) == 64 then {
+    // The extend and truncate ops below are no-ops when xlen==64,
+    // but appease the type-checker when xlen==32 (when this code is not executed!)
+    let o64      : bits(64) = zero_extend(o);
+    let v64      : bits(64) = zero_extend(v);
+    let new_satp : bits(64) = legalize_satp64(a, o64, v64);
+    truncate(new_satp, sizeof(xlen))
+  } else
+    internal_error(__FILE__, __LINE__, "Unsupported xlen" ^ string_of_int(sizeof(xlen)))
+}
+
+// ----------------
+// Fields of SATP
+
+// ASID is 9b in Sv32, 16b in Sv39/Sv48/Sv57: we use 16b for both
+// PRIVATE
+function satp_to_asid(sv_params : SV_Params, satp_val : xlenbits) -> bits(16) = {
+  // This extend op is a no-op when xlen==64, extends when xlen==32
+  let satp_64b : bits(64) = zero_extend(satp_val);
+  let mask_64b : bits(64) = zero_extend(ones(sv_params.satp_asid_size_bits));
```

Comment:
I don't think it makes sense for `satp` field sizes to be in `sv_params` because they only depend on XLEN, not the Sv39,Sv48,etc. mode. In other words this function should be:

```
function satp_to_asid(satp_val : xlenbits) -> bits(16) =
   if sizeof(xlen) == 32 then zero_extend(Mk_Satp32(satp_val)[Asid])
   else if sizeof(xlen) == 64 then Mk_Satp64(satp_val)[Asid]
   else internal_error(__FILE__, __LINE__, "Unsupported xlen" ^ string_of_int(sizeof(xlen)))
```

Reply:
Ok. Adopted this suggestion.